lib/eosknowledgeprivate/ekn-media-bin.c

```diff
@@ -0,0 +1,1462 @@
+
+  if (tags)
+    {
+      GstStructure *data = gst_structure_new (name, "tags", GST_TYPE_TAG_LIST, tags, NULL);
+      /* Post message on the bus for the main thread to pick it up */
+      gst_element_post_message (playbin,
+                                gst_message_new_application (GST_OBJECT (playbin),
+                                                             data));
+    }
+}
+
+static void
+on_playbin_audio_tags_changed (GstElement *play, gint stream_id, gpointer name)
+{
+  on_playbin_tags_changed (play, stream_id, name, "current-audio", "get-audio-tags");
+}
+
+static void
+on_playbin_video_tags_changed (GstElement *play, gint stream_id, gpointer name)
+{
+  on_playbin_tags_changed (play, stream_id, name, "current-video", "get-video-tags");
+}
+
+static void
+on_playbin_text_tags_changed (GstElement *play, gint stream_id, gpointer name)
+{
+  on_playbin_tags_changed (play, stream_id, name, "current-text", "get-text-tags");
+}
+
+static void
+ekn_media_bin_init_playbin (EknMediaBin *bin)
+{
+  bin->play = gst_element_factory_make ("playbin", "EknMediaBinPlayBin");
+
+  g_signal_connect (bin->play, "audio-tags-changed",
+                    G_CALLBACK (on_playbin_audio_tags_changed),
+                    "audio-tags-changed");
+  g_signal_connect (bin->play, "video-tags-changed",
+                    G_CALLBACK (on_playbin_video_tags_changed),
+                    "video-tags-changed");
+  g_signal_connect (bin->play, "text-tags-changed",
+                    G_CALLBACK (on_playbin_text_tags_changed),
+                    "text-tags-changed");
+
+  /* Setup volume */
+  /* NOTE: Bidirectional binding makes the app crash on X11 */
+  g_object_bind_property (bin->volume_button, "value",
+                          bin->play, "volume",
+                          G_BINDING_SYNC_CREATE);
+
+  /* Watch bus */
+  bin->bus = gst_pipeline_get_bus (GST_PIPELINE (bin->play));
+  gst_bus_add_watch (bin->bus, ekn_media_bin_bus_watch, bin);
+}
+
+/********************************* Public API *********************************/
+
+/**
+ * ekn_media_bin_new:
+ *
+ * Returns a new #EknMediaBin
+ *
+ */
+GtkWidget *
+ekn_media_bin_new ()
+{
+  return (GtkWidget*) g_object_new (EKN_TYPE_MEDIA_BIN, NULL);
+}
+
+/* The following macros are used to define generic getters and setters */
+
+#define EMB_DEFINE_GETTER(type, prop, retval) \
+type \
+ekn_media_bin_get_##prop (EknMediaBin *bin) \
+{ \
+  g_return_val_if_fail (EKN_IS_MEDIA_BIN (bin), retval); \
+  return bin->prop; \
+}
+
+/* The last argument is for custom code that will be added just before calling
+ * g_object_notify_by_pspec()
+ *
+ * You have access to EknMediaBin *bin and the property defined by prop
+ */
+#define EMB_DEFINE_SETTER(type, prop, PROP, code) \
+void \
+ekn_media_bin_set_##prop (EknMediaBin *bin, type prop) \
+{ \
+  g_return_if_fail (EKN_IS_MEDIA_BIN (bin)); \
+  if (bin->prop != prop) \
+    { \
+      bin->prop = prop; \
+      code \
```

Comment:
hmm now that I think more about it, we do not need them since this macro should not be used as a statement, its defining a function.